From 6ec9b6fb0fea706e4cc436b74e065cdad12df55c Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Thu, 12 Mar 2026 15:59:17 +0100
Subject: [PATCH] d2d1: Do not reference layer objects when recording to the
 command list.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
(cherry picked from commit 8818d2aca65b18ccd35b62c8d4030ac35c5fdba3)
---
 dlls/d2d1/command_list.c | 7 ++-----
 dlls/d2d1/d2d1_private.h | 2 +-
 dlls/d2d1/device.c       | 4 ++--
 3 files changed, 5 insertions(+), 8 deletions(-)

diff --git a/dlls/d2d1/command_list.c b/dlls/d2d1/command_list.c
index 0ca1bb9c0bf..b84eb74ba6f 100644
--- a/dlls/d2d1/command_list.c
+++ b/dlls/d2d1/command_list.c
@@ -117,7 +117,6 @@ struct d2d_command_push_layer
 {
     struct d2d_command c;
     D2D1_LAYER_PARAMETERS1 params;
-    ID2D1Layer *layer;
 };
 
 struct d2d_command_draw_line
@@ -477,7 +476,7 @@ static HRESULT STDMETHODCALLTYPE d2d_command_list_Stream(ID2D1CommandList *iface
             case D2D_COMMAND_PUSH_LAYER:
             {
                 const struct d2d_command_push_layer *c = data;
-                hr = ID2D1CommandSink_PushLayer(sink, &c->params, c->layer);
+                hr = ID2D1CommandSink_PushLayer(sink, &c->params, NULL);
                 break;
             }
             case D2D_COMMAND_POP_CLIP:
@@ -743,7 +742,7 @@ void d2d_command_list_pop_clip(struct d2d_command_list *command_list)
 }
 
 void d2d_command_list_push_layer(struct d2d_command_list *command_list, const struct d2d_device_context *context,
-        const D2D1_LAYER_PARAMETERS1 *params, ID2D1Layer *layer)
+        const D2D1_LAYER_PARAMETERS1 *params)
 {
     struct d2d_command_push_layer *command;
     ID2D1Brush *opacity_brush = NULL;
@@ -755,14 +754,12 @@ void d2d_command_list_push_layer(struct d2d_command_list *command_list, const st
         return;
     }
 
-    d2d_command_list_reference_object(command_list, layer);
     d2d_command_list_reference_object(command_list, params->geometricMask);
 
     command = d2d_command_list_require_space(command_list, sizeof(*command));
     command->c.op = D2D_COMMAND_PUSH_LAYER;
     command->params = *params;
     command->params.opacityBrush = opacity_brush;
-    command->layer = layer;
 }
 
 void d2d_command_list_pop_layer(struct d2d_command_list *command_list)
diff --git a/dlls/d2d1/d2d1_private.h b/dlls/d2d1/d2d1_private.h
index dcfdf6186fa..fd4a0f165fc 100644
--- a/dlls/d2d1/d2d1_private.h
+++ b/dlls/d2d1/d2d1_private.h
@@ -951,7 +951,7 @@ void d2d_command_list_fill_mesh(struct d2d_command_list *command_list, const str
 void d2d_command_list_fill_opacity_mask(struct d2d_command_list *command_list, const struct d2d_device_context *context,
         ID2D1Bitmap *bitmap, ID2D1Brush *orig_brush, const D2D1_RECT_F *dst_rect, const D2D1_RECT_F *src_rect);
 void d2d_command_list_push_layer(struct d2d_command_list *command_list, const struct d2d_device_context *context,
-        const D2D1_LAYER_PARAMETERS1 *params, ID2D1Layer *layer);
+        const D2D1_LAYER_PARAMETERS1 *params);
 void d2d_command_list_pop_layer(struct d2d_command_list *command_list);
 
 static inline BOOL d2d_array_reserve(void **elements, size_t *capacity, size_t count, size_t size)
diff --git a/dlls/d2d1/device.c b/dlls/d2d1/device.c
index 83b530db80a..5f4f7abfd85 100644
--- a/dlls/d2d1/device.c
+++ b/dlls/d2d1/device.c
@@ -1852,7 +1852,7 @@ static void STDMETHODCALLTYPE d2d_device_context_PushLayer(ID2D1DeviceContext6 *
 
         memcpy(&parameters, layer_parameters, sizeof(*layer_parameters));
         parameters.layerOptions = D2D1_LAYER_OPTIONS1_NONE;
-        d2d_command_list_push_layer(context->target.command_list, context, &parameters, layer);
+        d2d_command_list_push_layer(context->target.command_list, context, &parameters);
     }
 }
 
@@ -2810,7 +2810,7 @@ static void STDMETHODCALLTYPE d2d_device_context_ID2D1DeviceContext_PushLayer(ID
     FIXME("iface %p, layer_parameters %p, layer %p stub!\n", iface, layer_parameters, layer);
 
     if (context->target.type == D2D_TARGET_COMMAND_LIST)
-        d2d_command_list_push_layer(context->target.command_list, context, layer_parameters, layer);
+        d2d_command_list_push_layer(context->target.command_list, context, layer_parameters);
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_device_context_InvalidateEffectInputRectangle(ID2D1DeviceContext6 *iface,
